### PR TITLE
Add a SparseSet implementation

### PR DESCRIPTION
**Pull request by OneAvargeCoder193**

### CODE REVIEW

File: src/utils.zig [hunk 1]
```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
...
+
+		pub fn remove(self: *Self, allocator: NeverFailingAllocator, sparseId: idType) void {
+			if(!self.contains(sparseId)) return;
+
+			const denseId = self.sparse.items[sparseId];
+			if(denseId == noValue) return;
+
+			self.freeList.append(allocator, sparseId);
+
+			if(self.dense.items.len == 1) {
```
Comment (IntegratedQuantum) [inline src/utils.zig:1962]: This doesn't make sense. There is no reason it should behave different for when there is only one element.
You can safely remove this branch without causing any problems.

File: src/utils.zig [hunk 2]
```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
...
+				_ = self.dense.pop();
+			} else {
+				self.dense.items[denseId] = self.dense.pop();
+			}
+
+			self.sparse.items[sparseId] = noValue;
+			self.sparse.items[self.dense.items[denseId].id] = denseId;
+		}
+
+		pub fn get(self: *Self, id: idType) GetError!*T {
```
Comment (IntegratedQuantum) [inline src/utils.zig:1972]: In my opinion this should return an Optional.
An error in my mind signals an exceptional state, but I'd guess that you'd probably want to use this to check if the element exists, and that's what an optional is for.

File: src/utils.zig [hunk 3]
```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
...
+			const index = self.sparse.items[id];
+			if(index == noValue) return GetError.valueDoesntExist;
+			return &self.dense.items[index].value;
+		}
+	};
+}
+
+const SparseSetTest = struct {
+	var testingAllocator = main.heap.ErrorHandlingAllocator.init(std.testing.allocator);
+	var allocator = testingAllocator.allocator();
```
Comment (IntegratedQuantum) [inline src/utils.zig:1983]: We should probably have a wrapped testing allocator somewhere globally, please place this as `main.heap.testingAllocator`

File: src/utils.zig [hunk 4]
```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
...
+			self.sparse.deinit(allocator);
+			self.freeList.deinit(allocator);
+		}
+
+		pub fn contains(self: *Self, id: idType) bool {
+			return id < self.sparse.items.len and self.sparse.items[id] != noValue;
+		}
+
+		pub fn add(self: *Self, allocator: NeverFailingAllocator, value: T) idType {
+			const sparseId: idType = self.freeList.popOrNull() orelse @intCast(self.sparse.items.len);
```
Comment (IntegratedQuantum) [inline src/utils.zig:1941]: The sparse id, currently the return type, should be the input parameter, otherwise it's kind of useless for an ECS where we'd have multiple sparse sets for all the components, but only id per entity.

File: src/utils.zig [hunk 5]
```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
...
+
+	set.remove(SparseSetTest.allocator, firstId);
+
+	const newId = set.add(SparseSetTest.allocator, expectNew);
+
+	try std.testing.expectEqual(firstId, newId);
+
+	try std.testing.expectEqual((try set.get(secondId)).*, expectSecond);
+	try std.testing.expectEqual((try set.get(newId)).*, expectNew);
+}
```
Comment (IntegratedQuantum) [inline src/utils.zig:2027]: All the tests are doing almost the same thing, so I'd suggest to merge them.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]

File: src/utils.zig (hunk 1 above)
Comment (Argmaster) [reply src/utils.zig:1962]: If I pop when there is only 1 element then list has length of 0. If I try indexing at 0 wont it be angry at me?

File: src/utils.zig (hunk 5 above)
Comment (Argmaster) [reply src/utils.zig]: I don't agree, tests should be simple and should test specific edge cases. We added few more test cases testing few more edge cases. Test should follow arrange, act, assert (AAA) rule, which we do here, otherwise they get complicated and their failure is ambiguous. If we were to merge the tests, we would either neglect verifying some of the outputs, or we would have multiple of act-assert segments.

**Comment by IntegratedQuantum** [comment]
I thought about this some more, and I don't think that the SparseSet should contain the actual data, instead I think the data should be managed on the outside, to allow e.g. using different data structures, e.g. all entities would have health and position data, so both of them could share the same dense index, yet at the same time they should probably be in separate Lists to increase cache efficiency.

Furthermore this would allow using vector-aligned and overallocated lists for some components to allow vectorizing the system operations. Bit-packed lists might also be useful.

File: src/utils.zig (hunk 5 above)
Comment (IntegratedQuantum) [reply src/utils.zig]: Well I was just saying this because before the last test was basically doing the same thing as the two previous tests, but the new tests are a lot better in that regard.

**Comment by Argmaster** [comment]
You are already optimizing something that neither exist nor was proven to be performance bottleneck. This will greatly complicate the system logic with questionable performance gains. As a first proof of concept we should do the simple, fairly optimized way everybody does, then if it is proven to be too slow, we can explore different possibilities and if benchmarks prove there is performance benefit we should go for it.

**Comment by Argmaster** [comment]
I need to see the code, can't do that now. However, this would no longer be a sparse set and I don't understand what and how do you want it to be implemented.

File: src/utils.zig [hunk 6]
```diff
@@ -1907,6 +1907,155 @@ test "read/write mixed" {
...
+		const Index = enum(IdType) {
+			noValue = std.math.maxInt(IdType),
+			_,
+		};
+
+		const Self = @This();
+
+		dense: main.ListUnmanaged(T) = .{},
+		reverseIds: main.ListUnmanaged(IdType) = .{},
+		sparse: main.ListUnmanaged(Index) = .{},
```
Comment (IntegratedQuantum) [inline src/utils.zig:1924]: Maybe we could call them denseToSparseIndex and sparseToDenseIndex?

File: src/utils.zig [hunk 7]
```diff
@@ -1907,6 +1907,155 @@ test "read/write mixed" {
...
+		}
+
+		pub fn set(self: *Self, allocator: NeverFailingAllocator, id: IdType, value: T) void {
+			if(id >= self.sparse.items.len) {
+				self.sparse.appendNTimes(allocator, .noValue, id - self.sparse.items.len + 1);
+			}
+
+			std.debug.assert(self.sparse.items[id] == .noValue);
+
+			const denseId: Index = @enumFromInt(self.dense.items.len);
```
Comment (IntegratedQuantum) [inline src/utils.zig:1943]: There is still a problem here: What if `self.dense.items.len` is `.noValue`?

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]

**Comment by IntegratedQuantum** [comment]
> I thought about this some more, and I don't think that the SparseSet should contain the actual data, instead I think the data should be managed on the outside, to allow e.g. using different data structures, e.g. all entities would have health and position data, so both of them could share the same dense index, yet at the same time they should probably be in separate Lists to increase cache efficiency.

Doesn't need to be done right now. When we get to this, it could be solved with a MultiArrayList variant of the SparseSet.

File: src/utils.zig (hunk 6 above)
Comment (Argmaster) [reply src/utils.zig:1924]: There is also no tests covering that while there should be one since this is a very important edge case.

File: src/utils.zig (hunk 7 above)
Comment (Argmaster) [reply src/utils.zig:1943]: There is also no tests covering that while there should be one since this is a very important edge case.

**Comment by OneAvargeCoder193** [comment]
Alright I now made it so that when the denseId would be equal to noValue, it returns without allocating anything. I also added a test for it.

**Comment by Argmaster** [comment]
IMO it should explicitly report an error, otherwise caller would have to every time check if the insertion was successful / understand internal implementation of SparseSet and and check for `noValue` before calling set.

File: src/utils/heap.zig [hunk 8]
```diff
@@ -1,10 +1,15 @@
...
 
 const main = @import("main");
 
 var testingErrorHandlingAllocator = ErrorHandlingAllocator.init(std.testing.allocator);
-pub var testingAllocator = testingErrorHandlingAllocator.allocator();
+pub var testingAllocator = blk: {
+	if(!builtin.is_test) @compileError("testing allocator used when not testing");
+
+	break :blk testingErrorHandlingAllocator.allocator();
+};
```
Comment (IntegratedQuantum) [inline src/utils/heap.zig:12]: This code already exists for std.testing.allocator, why do we need it here again?

File: src/utils/heap.zig [hunk 9]
```diff
@@ -3,6 +3,9 @@ const Allocator = std.mem.Allocator;
 
 const main = @import("main");
 
+var testingErrorHandlingAllocator = ErrorHandlingAllocator.init(std.testing.allocator);
+pub var testingAllocator = testingErrorHandlingAllocator.allocator();
```
Comment (IntegratedQuantum) [inline src/utils/heap.zig:7]: Should be const

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]

File: src/utils.zig [hunk 10]
```diff
@@ -1907,6 +1907,174 @@ test "read/write mixed" {
 	try std.testing.expect(reader.remaining.len == 0);
 }
 
+pub fn SparseSet(comptime T: type, comptime IdType: type) type { // MARK: SparseSet
+	std.debug.assert(@typeInfo(IdType) == .int);
+	std.debug.assert(@typeInfo(IdType).int.signedness == .unsigned);
+
+	return struct {
+		const Index = enum(IdType) {
```
Comment (IntegratedQuantum) [inline src/utils.zig:1915]: I'm not really happy with this. In my opinion `IdType` should be an enum with a `noValue` variant to begin with.
This would also remove the need fo error handling in the `set` function, since we can assume that the caller handled the `noValue` case. And also this gives some better type safety to the index type.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]
Given the recent proposal from @Argmaster on discord, I think it might be better to wait before adding this, until we actually know where and how we need it.

File: src/utils.zig (hunk 10 above)
Comment (OneAvargeCoder193) [reply src/utils.zig:1915]: ? the noValue is needed for the set function to make sure you dont accidentally set it over itself

Comment (OneAvargeCoder193) [reply src/utils.zig:1915]: in my opinion, needing to pass an enum with a noValue into the sparseset overcomplicates the SparseSet call.
We currently just do SparseSet(V, K)
but now we will need to do
SparseSet(V, enum(K) {
    noValue = std.math.maxInt(K),
    _,
})

Comment (IntegratedQuantum) [reply src/utils.zig:1915]: The point is that the enum would also be used for the Entity itself, which would be the right approach anyways in my opinion, enums are just more type safe.

File: src/utils.zig [hunk 11]
```diff
@@ -1907,6 +1907,206 @@ test "read/write mixed" {
...
+		pub fn contains(self: *Self, id: IdType) bool {
+			return @intFromEnum(id) < self.sparseToDenseIndex.items.len and self.sparseToDenseIndex.items[@intFromEnum(id)] != .noValue;
+		}
+
+		pub fn set(self: *Self, allocator: NeverFailingAllocator, id: IdType, value: T) !void {
+			const denseId: IdType = @enumFromInt(self.dense.items.len);
+
+			if(denseId == .noValue) {
+				return error.TooMany;
+			}
```
Comment (IntegratedQuantum) [inline src/utils.zig:1933]: Now that the parent type has the `.noValue` thing we don't need this check anymore.
Instead an `assert` that the input `id` is not `.noValue` should be enough.

File: src/utils.zig [hunk 12]
```diff
@@ -1907,6 +1907,206 @@ test "read/write mixed" {
 	try std.testing.expect(reader.remaining.len == 0);
 }
 
+pub fn SparseSet(comptime T: type, comptime IdType: type) type { // MARK: SparseSet
```
Comment (IntegratedQuantum) [inline src/utils.zig:1910]: Please assert that the `noValue` field has the maximum integer value of the backing enum of the IdType

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]

File: src/utils.zig (hunk 11 above)
Comment (OneAvargeCoder193) [reply src/utils.zig:1933]: denseId isn't the input though. In fact, there should never be a noValue when you index into sparseToDenseIndex, this was the main reason i decided to create the type inside SparseSet.

Comment (OneAvargeCoder193) [reply src/utils.zig:1933]: We still need this check because denseId isn't the id that is passed into the function. I wanted that to be a basic integer since sparseId never has a noValue index. noValue is only used as a value inside sparseSet

Comment (IntegratedQuantum) [reply src/utils.zig:1933]: > noValue is only used as a value inside sparseSet

Yes, but I think it makes sense to have a noValue value for the entity index as well, it serves as a limit for the number of entities. Without this the limit is not quite obvious from the outside, and ideally you'd check the entity limit long before you add anything into a SparseSet, as at that point it's usually too late and you need to potentially clean up other components that didn't cause an error yet.

Comment (Argmaster) [reply src/utils.zig:1933]: Can we have DenseId generic type that encapsulates this `noValue` behavior, basically appropriate enum, instead of it being contract specified by comment and assertions? Dense list will never grow bigger than sparse list is and sparse list limited by number of IDs

Comment (IntegratedQuantum) [reply src/utils.zig:1933]: > Can we have DenseId generic type that encapsulates this noValue behavior, basically appropriate enum, instead of it being contract specified by comment and assertions?

Yes that is probably a good idea. We will still need those assertions though to prevent you from passing the wrong thing.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]
Please fix the formatting, then we are good to go here

**Comment by OneAvargeCoder193** [comment]
done

**Review by IntegratedQuantum — APPROVED** [review]
